pkg/controllers/provisioningv2/rke2/machineprovision/controller.go

```diff
@@ -220,6 +251,53 @@ func (h *handler) OnRemove(_ string, obj runtime.Object) (runtime.Object, error)
 		return nil, err
 	}
 
+	// In the event we are removing an etcd node (as indicated by the etcd-role label on the node), we must safely remove the etcd node from the cluster before allowing machine deprovisioning
+	if val, ok := d.Map("metadata", "labels")["rke.cattle.io/etcd-role"]; ok && val.(string) == "true" {
+		// we need to block removal until our the v1 node that corresponds has been removed
+		clusterName, _ := d.Map("metadata", "labels")[CapiMachineLabel].(string)
+		if clusterName == "" {
+			return obj, fmt.Errorf("error retrieving the clustername for etcd node, label key %s does not appear to exist for dynamic machine %s", CapiMachineLabel, key)
+		}
+		cluster, err := h.rancherClusterCache.Get(d.String("metadata", "namespace"), clusterName)
+		if err != nil && !apierror.IsNotFound(err) {
+			return obj, err
+		}
+		if apierror.IsNotFound(err) || !cluster.DeletionTimestamp.IsZero() {
+			return h.doRemove(obj)
+		}
+
+		machine, err := h.getMachine(obj)
+		if err != nil {
+			return obj, err
+		}
+
+		if machine.Status.NodeRef == nil {
+			// Machine noderef is nil, we should just allow deletion.
+			logrus.Debugf("[MachineProvision] There was no associated K8s node with this etcd dynamicmachine %s. proceeding with deletion", key)
+			return h.doRemove(obj)
+		}
+
+		restConfig, err := h.kubeconfigManager.GetRESTConfig(cluster, cluster.Status)
+		if err != nil {
+			return obj, err
+		}
+
+		removed, err := etcdmgmt.SafeRemove(restConfig, v2provruntime.GetRuntimeCommand(cluster.Spec.KubernetesVersion), machine.Status.NodeRef.Name)
+
+		if err != nil || !removed {
```

Comment:
I think these cases should be separated. If there really was an error, return it and let the controller log it and retry.